Samples/NetworkMediaStreamer/RTSPServer/src/RTSPAuthProvider.cpp

```diff
@@ -0,0 +1,228 @@
+        std::string authString;
+        if ((m_authType == AuthType::Both) || (m_authType == AuthType::Basic))
+        {
+            authString = "WWW-Authenticate: Basic realm=\"BeyondTheWall\", charset=\"UTF-8\"\r\n";
+        }
+        auto buf = Cryptography::CryptographicBuffer::GenerateRandom(24);
+        *((uint64_t*)(buf.data() + 16)) = MFGetSystemTime();
+        auto nonce = Cryptography::CryptographicBuffer::EncodeToHexString(buf);
+        if ((m_authType == AuthType::Both) || (m_authType == AuthType::Digest))
+        {
+            authString += "WWW-Authenticate: Digest realm=\"BeyondTheWall\", nonce=\"" + winrt::to_string(nonce) + "\", algorithm=SHA-256, charset=\"UTF-8\", stale=FALSE\r\n";
+            authString += "WWW-Authenticate: Digest realm=\"BeyondTheWall\", nonce=\"" + winrt::to_string(nonce) + "\", algorithm=MD5, charset=\"UTF-8\", stale=FALSE\r\n";
+        }
+
+        *pAuthSessionMessage = reinterpret_cast<HSTRING> (winrt::detach_abi(winrt::to_hstring(authString)));
+        return S_OK;
+    }HRESULT_EXCEPTION_BOUNDARY_FUNC
+
+
+    STDMETHODIMP Authorize(LPCWSTR authResp, LPCWSTR authSesMsg, LPCWSTR mthd) override try
+    {
+        winrt::check_pointer(authResp);
+        winrt::check_pointer(authSesMsg);
+        winrt::check_pointer(mthd);
+        bool result = true;
+        auto authResponse = winrt::to_string(authResp);
+        auto authSessionMessage = winrt::to_string(authSesMsg);
+        auto method = winrt::to_string(mthd);
+        std::vector<std::string> paramsResponce = { "Authorization: ","realm=","nonce=", "username=", "response=", "algorithm=", "uri=" };
+        std::vector<std::string> paramsSent = { "realm=","nonce=" };
+        std::map<std::string, std::string> paramMapResponse, paramMap;
+        for (auto param : paramsResponce)
+        {
+            paramMapResponse[param] = GetValueForParam(authResponse, param);
+        }
+        for (auto param : paramsSent)
+        {
+            paramMap[param] = GetValueForParam(authSessionMessage, param);
+            if (!((paramMapResponse[param].empty()) || (paramMap[param].empty())))
+            {
+                result &= (paramMapResponse[param] == paramMap[param]);
+            }
+        }
+
+        if (paramMapResponse["Authorization: "] == "Digest")
+        {
+            CryptographicHash hash(nullptr);
+            if ((paramMapResponse["algorithm="] == "MD5") || (paramMapResponse["response="].size() < 64))
+            {
+                hash = m_hashMD5;
+            }
+            else
+            {
+                hash = m_hashSHA256;
+            }
+            auto username = paramMapResponse["username="];
+            std::string password;
+            auto vault = PasswordVault();
+            
+            PasswordCredential cred;
+            if (!username.empty())
+            {
+                cred = vault.Retrieve(m_resourceName, winrt::to_hstring(username));

```

Comment:
>    cred = vault.Retrieve(m_resourceName, winrt::to_hstring(username)); [](start = 12, length = 71)

no "ugly-hack" here to handle gracefully?

Reply:
i.e. your else clause at line 103 won't trigger ever currently since Retrieve will throw in case of failure to retrieve and will be handled in your catch(...) below

---
In reply to: [492986135](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r492986135) [](ancestors = 492986135)

Reply:
oops... looks like the try catch  got cleaned up during some refactoring ....i shall add it back

---
In reply to: [492989430](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r492989430) [](ancestors = 492989430,492986135)